spirv_msl.cpp

```diff
@@ -3016,6 +3327,121 @@ void CompilerMSL::emit_instruction(const Instruction &instruction)
 		break;
 	}
 
+	case OpInBoundsAccessChain:
+	case OpAccessChain:
+	case OpPtrAccessChain:
+		if (get_execution_model() == ExecutionModelTessellationControl)
+		{
+			// If this is a per-vertex output, remap it to the I/O array buffer.
+			auto *var = maybe_get<SPIRVariable>(ops[2]);
+			BuiltIn bi_type = BuiltIn(get_decoration(ops[2], DecorationBuiltIn));
+			if (var && (var->storage == StorageClassInput || var->storage == StorageClassOutput) &&
+			    !has_decoration(ops[2], DecorationPatch) &&
+			    (!is_builtin_variable(*var) || bi_type == BuiltInPosition || bi_type == BuiltInPointSize ||
+			     bi_type == BuiltInClipDistance || bi_type == BuiltInCullDistance ||
+			     get_variable_data_type(*var).basetype == SPIRType::Struct))
+			{
+				uint32_t length = instruction.length;
+				AccessChainMeta meta;
+				std::vector<uint32_t> indices;
+				uint32_t next_id = ir.increase_bound_by(2);
+
+				indices.reserve(length - 3 + 1);
+				uint32_t type_id = next_id++;
+				SPIRType new_uint_type;
+				new_uint_type.basetype = SPIRType::UInt;
+				new_uint_type.width = 32;
+				set<SPIRType>(type_id, new_uint_type);
+
+				indices.push_back(ops[3]);
+
+				uint32_t const_mbr_id = next_id++;
+				uint32_t index = get_extended_decoration(ops[2], SPIRVCrossDecorationInterfaceMemberIndex);
+				if (var->storage == StorageClassInput ||
+				    has_decoration(get_variable_element_type(*var).self, DecorationBlock))
+				{
+					uint32_t i = 4;
+					if (index == uint32_t(-1))
+					{
+						// Maybe this is a struct type in the input class, in which case
+						// we put it as a decoration on the corresponding member.
+						index =
+						    get_extended_member_decoration(ops[2], ops[4], SPIRVCrossDecorationInterfaceMemberIndex);
+						assert(index != uint32_t(-1));
+						i++;
+					}
+					// In this case, we flattened structures and arrays, so now we have to
+					// combine the following indices. If we encounter a non-constant index,
+					// we're hosed.
+					auto *type = &get_variable_element_type(*var);
+					for (; i < length; ++i)
+					{
+						if (!is_array(*type) && !is_matrix(*type) && type->basetype != SPIRType::Struct)
+							break;
+
+						auto &c = get_constant(ops[i]);
+						index += c.scalar();
+						if (type->parent_type)
+							type = &get<SPIRType>(type->parent_type);
+					}
+					set<SPIRConstant>(const_mbr_id, type_id, index, false);
+					indices.push_back(const_mbr_id);
+
+					indices.insert(indices.end(), &ops[i], &ops[length]);
```

Comment:
This is broken on a test shader I have.

Reply:
```
#version 450

struct Boo
{
        vec3 a;
        vec3 b;
};

layout(vertices = 4) out;
layout(location = 0) out Boo vVertex[];
layout(location = 0) in Boo vInput[];

void main()
{
        vVertex[gl_InvocationID] = vInput[gl_InvocationID];
        gl_TessLevelOuter[0] = 1.0;
        gl_TessLevelOuter[1] = 2.0;
        gl_TessLevelOuter[2] = 3.0;
        gl_TessLevelOuter[3] = 4.0;
        gl_TessLevelInner[0] = 1.0;
        gl_TessLevelInner[1] = 2.0;
}
```
i is 5 here, yet length is 4. Also, I think dereferencing out of bounds is not technically valid, please change it to ops + i and ops + length to avoid asan/valgrind/debug STL complaining.

Reply:
I tried hacking in this patch:
```
@@ -3368,7 +3368,7 @@ void CompilerMSL::emit_instruction(const Instruction &instruction)
                                                index =
                                                    get_extended_member_decoration(ops[2], ops[4], SPIRVCrossDecorationInterfaceMemberIndex);
                                                assert(index != uint32_t(-1));
-                                               i++;
+                                               //i++;
                                        }
```
That lets me compile, but it does expose another issue I think. The code-gen is:
```
    gl_out[gl_InvocationID].vVertex = gl_in[gl_InvocationID].Boo_a;
```
I think we need to do the unflattening step which we currently do for I/O structs in normal vertex shading to be able to access them as whole structs.

Reply:
After this is all done, I've half a mind to file a radar requesting support for composite types other than vectors in stage-in structs. This flattening/unflattening stuff is getting out of hand.

Reply:
Yup, it's getting nightmarish. It's painful enough without tesc. Feel free to just drop support for composites in tesc if it gets out of hand as long as we can detect it and just fail gracefully.

Reply:
OK, I finally got this figured out. (I think.)